### DISCUSSION TITLE
Can't spawn a new tab when called from within tmux?

### DISCUSSION THREAD

**Discussion by 097115** [kovidgoyal/kitty#6017] [Q&A] [answered]
I have a kitty window with one tab running a tmux session. Now, from within that session in a shell prompt I issue the following command: `kitty @ launch --type=tab --tab-title "My Tab"` -- and it just freezes?

If I issue the same command from a regular kitty window, then everything works ok, a new tab is spawned, and a tab bar appears.

What I'd like to achieve: while being in my tmux session launch a script, which would open a new tab in the same kitty window with w3m showing some user-specified URL there. What am I missing?

Thank you.

**Comment by kovidgoyal** [comment, marked as answer]
You are missing that tmux filters out escape codes it doesnt understand.
So use a socket for remote control.

**Reply by page-down** [reply]
> ... tmux filters out escape codes ...

I think the following can be implemented in Go for `kitten @`.
https://github.com/kovidgoyal/kitty/commit/b3fcb53625920065a35c7fcae230344fadf83275

Or wait until ssh kitten needs to be ported to the new `kitten`.

> ... the KITTY_LISTEN_ON env var is set to the correct socket ...

In this scenario, if the user attaches to a running tmux instance started by another kitty instance,
this environment variable in the shell is possibly inaccurate.

**Comment by 097115** [comment]
It works, thanks!

The only obstacle is that kitty doesn't find the socket given its template: if I have `listen_on unix:/tmp/mykitty` set in kitty.conf, then the full socket name will something like `/tmp/mykitty-876826`, and in the command line I have to address it fully, not just `kitty @ --to unix:/tmp/mykitty [...]`. But that's solvable via `kitty @ --to unix:$(ls /tmp/mykitty*) [...]`.

**Reply by page-down** [reply]
> if I have `listen_on unix:/tmp/mykitty` set in kitty.conf ...

You can run many `kitty`s at the same time, and all will read the same configuration file.
They cannot listen to the same UNIX socket, so it will automatically append the PID.

**Reply by kovidgoyal** [reply]
> It works, thanks!
> 
> The only obstacle is that kitty doesn't find the socket given its template: if I have `listen_on unix:/tmp/mykitty` set in kitty.conf, then the full socket name will something like `/tmp/mykitty-876826`, and in the command line I have to address it fully, not just `kitty @ --to unix:/tmp/mykitty [...]`. But that's solvable via `kitty @ --to unix:$(ls /tmp/mykitty*) [...]`.

You dont need to do all that the KITTY_LISTEN_ON env var is set to the correct socket automatically and kitty @ will use that env var automatically.

**Reply by SichangHe** [reply]
This has changed, right? `KITTY_LISTEN_ON` is not there when I check now, but the command still works.

**Comment by 097115** [comment]
> KITTY_LISTEN_ON env var is set 

It seems it gets unset in tmux... Is it possible?

I launch kitty, `echo $KITTY_LISTEN_ON` returns the correct socket. Then I start tmux in this very same kitty window, and `echo $KITTY_LISTEN_ON` returns nothing.

`kitty 0.26.5`

**Comment by kovidgoyal** [comment]
No clue, I dont touch tmux.

**Comment by kovidgoyal** [comment]
On Tue, Feb 14, 2023 at 04:33:27AM -0800, page-down wrote:
> > ... tmux filters out escape codes ...
> 
> I think the following can be implemented in Go for `kitten @`.
> https://github.com/kovidgoyal/kitty/commit/b3fcb53625920065a35c7fcae230344fadf83275
> 
> Or wait until ssh kitten needs to be ported to the new `kitten`.

No, not really, remote control requires two way communication in general.
tmux passthrough is insufficient.

> 
> > ... the KITTY_LISTEN_ON env var is set to the correct socket ...
> 
> In this scenario, if the user attaches to a running tmux instance started by another kitty instance,
> this environment variable in the shell is possibly inaccurate.

Yeah, that's true, so use kitty --single-instance to run all your
kitties. Or better, dont use tmux :)

**Comment by SichangHe** [comment]
This also does not work when called from within Neovim terminal, btw. Tried that to preview images from within Neovim.
